Needs to snap point too when CadDockWidget is enabled. Fixes #18138

lbartoletti opened the pull request:
## Description
The PR corrects the Z support when the CadDock is activated. It only adds a test, which calls the snapPoint function if a snapped point is a vertex like when the CadDock is not enabled. 

This fixes the bug #18138

I also updated mSnappedToVertex which I thought was never modified in this case.

@wonder-sk @3nids is it ok for you ?

cc @ponceta @pblottiere 

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [X] Commit messages are descriptive and explain the rationale for changes
- [X] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [X] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [X] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

lbartoletti committed:
Needs to snap point too when CadDockWidget is enabled. Fixes #18138

3nids commented on src/gui/qgsmaptooladvanceddigitizing.cpp:
```diff
@@ -34,6 +34,11 @@ void QgsMapToolAdvancedDigitizing::canvasPressEvent( QgsMapMouseEvent *e )
 
+    if ( mCadDockWidget->mapPointMatch().hasVertex() )
+    {
+      e->snapPoint();
```
If I understand correctly, the idea is to force snapping to a point (and return to original point if failed).
can you explain the reasoning? Mainly which scenario.

lbartoletti replied:
When you activate the CadDock, it only returns the 2D point, by doing this if the vertex is identified, it can return a 3D point

3nids commented on src/gui/qgsadvanceddigitizingdockwidget.cpp:
```diff
@@ -555,6 +555,8 @@ bool QgsAdvancedDigitizingDockWidget::applyConstraints( QgsMapMouseEvent *e )
 
   mSnapMatch = context.snappingUtils->snapToMap( point );
 
+  mSnappedToVertex = mSnapMatch.hasVertex();
```
wouldn't it be safer to remove this unused member and rather adapt the method `snappedToVertex()` to return `mSnapMatch.isValid() && mSnapMatch.hasVertex()` ?

lbartoletti replied:
@3nids right. Done. Thanks

lbartoletti committed:
remove useless variable mSnappedToVertex

lbartoletti commented:
@3nids travis errors are not related

lbartoletti commented:
Screencast

Z default: 34

Snap with normal mode: works, with advanced widget: nope
![qgis_snap_z_3 6](https://user-images.githubusercontent.com/7521540/54125174-8f861d80-4404-11e9-8d5d-19e5a6c0748c.gif)

Fixed version
![qgis_snap_z_master](https://user-images.githubusercontent.com/7521540/54125172-8f861d80-4404-11e9-8efb-eee83eb37453.gif)

lbartoletti commented:
@3nids Can you merge it, please?

3nids commented:
@wonder-sk can we get your eyes on this one? Cheers

wonder-sk commented:
I have some trouble understanding the logic here and how exactly the 2D->3D conversion happens...

Shouldn't we do the `e->snapPoint()` inside applyConstraints() call? That would also avoid the duplicate call in mouse press / mouse release handler. And the same thing should be probably applied in mouse move handler too - or not?

Can you please add a comment explaining why the extra `e->snapPoint()` is in the code and what does it do? Otherwise I believe someone could just delete it thinking it does not have any good purpose.

I see you have discovered that snappedToVertex() has been returning false all the time... That means the logic in the advanced cad canvas item has changed too - does it work correctly now?

lbartoletti committed:
apply martin's review

lbartoletti commented:
@wonder-sk @3nids Thanks. I updated the code with a comments into applyConstraints(). I added a condition to avoid that the snapping does not exceed the constraints defined by the user

> I see you have discovered that snappedToVertex() has been returning false all the time... That means the logic in the advanced cad canvas item has changed too - does it work correctly now?

AFAIK yes

lbartoletti commented:
@3nids @wonder-sk Thanks.
Can it be backported?

lbartoletti commented:
With our magic backport bot ;)

3nids commented:
@lbartoletti done!
